xmake/rules/platform/windows/idl/xmake.lua

```diff
@@ -39,26 +67,130 @@ rule("platform.windows.idl")
-        batchcmds:vrunv(midl.program, flags, {envs = msvc:runenvs()})
+        -- use this way to set depend to avoid generating files multiple times
+        depend.on_changed(function() 
+            progress.show(opt.progress, "${color.build.object}generating.idl %s", sourcefile)
+            os.vrunv(midl.program, flags, { envs = msvc:runenvs() })
+        end, { files = sourcefile, dependfile = path.join(autogendir, path.basename(sourcefile) .. ".idl.d") })
         
-        local iid_file = path.join(autogendir, name .. "_i.c")
-        local objectfile = target:objectfile(iid_file)
-        table.insert(target:objectfiles(), objectfile)
+        --batchcmds:show_progress(opt.progress, "${color.build.object}compiling.idl %s", sourcefile)
+        --batchcmds:vrunv(midl.program, flags, {envs = msvc:runenvs()})
+
+    end)
+    --[[
+        we don't have a way to detect which files midl.exe has generated and os.exists
+        does not work in before_buildcmd_file because in the invokation of xmake
+        the files might not have been generated yet from batchcmds, therefore
+        the files are compiled and checked during the buildcmd as _i, _p, _c, _s
+        might not exists depending on the idl file
+    ]]
+    on_buildcmd_file(function (target, batchcmds, sourcefile, opt)
+        import("core.project.depend")
+
+        local name = path.basename(sourcefile)
+        local autogendir = path.join(target:autogendir(), "platform/windows/idl")
+
+        -- we don't have a way to detect which midl files are generated
+
+        local icfile = path.join(autogendir, name .. "_i.c")
+        local icobj = target:objectfile(icfile)
+
+        local scfile = path.join(autogendir, name .. "_s.c")
+        local scobj = target:objectfile(scfile)
+
+        local ccfile = path.join(autogendir, name .. "_c.c")
+        local ccobj = target:objectfile(ccfile)
+
+        local pcfile = path.join(autogendir, name .. "_p.c")
+        local pcobj = target:objectfile(pcfile)
+
+        local fileconfig = target:fileconfig(sourcefile)
+        local enable_proxy = true
+        if fileconfig then
+            if fileconfig.proxy ~= nil then
+                enable_proxy = fileconfig.proxy
+            end
+        end
+
+        -- compile c files
+        local configs = {includedirs = autogendir, languages = "c89"}
+        
+        if os.exists(icfile) then
+            table.insert(target:objectfiles(), icobj)
+            depend.on_changed(function()
+                batchcmds:show_progress(opt.progress, "${color.build.object}compiling.$(mode) %s", icfile)
+                batchcmds:compile(icfile, icobj, {sourcekind = "cxx", configs = configs})                 
+            end, { files = {icobj}, changed = target:is_rebuilt() })
+        end
+
+        if os.exists(scfile) then
+            table.insert(target:objectfiles(), scobj)
+            depend.on_changed(function()
```

Comment:
we should not use depend.on_changed in `buildcmx_xxx` callbacks.

see https://github.com/xmake-io/xmake/blob/fd129538be29cabad2c2650f9b837ab1afc28a1e/xmake/rules/lex_yacc/lex/xmake.lua#L47-L49

Reply:
in the case of the before_buildcmd_file, which file should I use as the set_depmtime? the files that midl generates (like _i.c, .h, _c.c) might not be generated depending on the configuration (like one configuration only generates a .tlb, one only generates the .h and _i.c and so on)